### PR TITLE
Dimensional_oM: Creation of the matter element IElementM

### LINKED ISSUE #729: Dimensional_oM: Matter Element, IElementM
#### Description:
Due partly to a rising need from the ClimateEmergency_Toolkit to be able to Query things related to the material of things, we should add a `IElementM` which handles things related to matter.

This would get a accompanying Matter_Engine which would be to the Physical_Engine what the Spatial_Engine is to the Geometry_Engine, in that it would allow a unified handling of elements with matter (physical objects) and their materials.

The main questions which `IElement`s will work with are:

1. How much matter is there?
2. What kind of matter am I?

#### Extension Methods:
The methods which would be required by the interface at a compliance level for functionality
- `SolidVolume(IElementM)`
- `MaterialComposition(IElementM)`

#### MaterialComposition
where `MaterialComposition()` would return a new class `MaterialComposition` which would contain:
- `List<Physical.Material>` Materials
- `List<double>` Ratios

where `Physical.Material` is composed of:
- `List<IMaterialFragment>` Properties

The `IMaterialFragement` means that each discipline can have a separate definition of what a material is

The Matter_Engine would largely create unified methods to handle this tree of properties which may or may not be present in a transparent manner.

A note is that BHoM does not force any kind of structure to the format of a material and a Physical.Material can contain multiple material definitions through `MaterialFragments` for different disciplines. Which is a good reason to create these overarching methods which would try and scrape the `MaterialFragments` of the useful information while also maintaining transparency about how good the information is. That is by attempting to get the information in different manners by attempting to query the properties and logging by which degree of accuracy it was attained.

#### Climate Emergency
The main things I've seen obstructing workflow in the ClimateEmergency_Toolkit has been a difficulty in obtaining the Volume and a need to define which material to use in each method. This will allow querying the SolidVolume and getting the assigned materials and their proportions, which may be a ClimateEmergency definition or queried from a library based on the other fragments. Would love for some input from ClimateEmergency folks on which specific features would be needed/helpful.

#### Code:
WIP branches for both oM and Engine: `Matter_Enigne-#-Creation-of-Matter_Engine` for those who are interested in seeing some examples of implementation. (note that I am aware that some of the methods do return wrong answers, but it's made to test the system not the results)

FYI @al-fisher @IsakNaslundBh @pawelbaran @michaelhoehn @kayleighhoude @FraserGreenroyd

### PR DESCRIPTION
**Pull request by kThorsager**
### Issues addressed by this PR
Closes #729 

### Test files
see https://github.com/BHoM/BHoM_Engine/pull/1550
### Changelog
- Creation of `MaterialComposition` which stores a collection of `Material`s and their ratios
- Set `Physical.IFramingElement` as `IElementM`
- Set `Structure.Bar` as `IElementM`

### CODE REVIEW

File: Physical_oM/Materials/MaterialComposition.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+using System.Threading.Tasks;
+
+using BH.oM.Base;
+using BH.oM.Geometry;
+using BH.oM.Quantities.Attributes;
+using System.ComponentModel;
+
+namespace BH.oM.Physical.Materials
+{
+    [Description("An object which defines a collection of materials and their ratios.")]
```
Comment (al-fisher) [inline Physical_oM/Materials/MaterialComposition.cs]:
```suggestion
    [Description("Defines the make up of an object through a list of Materials as relative proportions of the total solid volume." +
                        "\nThere must be the same number of items in both lists, assigning a single Ratio for each Material.")]
```

File: Physical_oM/Materials/MaterialComposition.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("The materials in the MaterialComposition, their order in the list is related to the order of the ratios.")]
+        public IReadOnlyList<Material> Materials { get; } = new List<Material>();
+
+        [Ratio]
+        [Description("The ratios of the Materials, the sum of which equals one")]
```
Comment (al-fisher) [inline Physical_oM/Materials/MaterialComposition.cs]:
```suggestion
        [Description("The list of Material volumetric Ratios. The sum of which must equate to one.")]
```

File: Physical_oM/Materials/MaterialComposition.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+namespace BH.oM.Physical.Materials
+{
+    [Description("An object which defines a collection of materials and their ratios.")]
+    public class MaterialComposition : BHoMObject, IPhysical, IImmutable
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("The materials in the MaterialComposition, their order in the list is related to the order of the ratios.")]
```
Comment (al-fisher) [inline Physical_oM/Materials/MaterialComposition.cs]:
```suggestion
        [Description("The Materials that form an object's make up, the order of which corresponds to the order of the Ratios.")]
```

File: Dimensional_oM/IElementM.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Base;
+using System.ComponentModel;
+
+namespace BH.oM.Dimensional
+{
+    [Description("Enables Mass based operations to be performed on elements with materiality\n" +
+                 "Ensures representation of the material composition of objects including the type and volume of matter forming the object's make up\n" +
```
Comment (al-fisher) [inline Dimensional_oM/IElementM.cs]:
```suggestion
    [Description("Enables Mass based operations to be performed on elements with materiality\n" +
                 "Ensures the material composition of a physical object is represented, defined as proportions of discrete types of material forming the object's total solid volume.\n" +
```

**Comment by al-fisher** [comment]
I have fixed the conflicts. Were caused simply by applying Spatial and Matter interfaces to the Physical objects `IFramingElement` and `ISurface` in two separate PRs.

@kThorsager @IsakNaslundBh - can you please do simple recheck and approval - all is as expected as my approval has been dismissed. 😄 Thanks

**Comment by kThorsager** [comment]
Can't approve, but have rechecked and am happy

**Pull request merged**